enclave-runtime/src/lib.rs

```diff
@@ -806,6 +809,60 @@ where
 	Ok(nonce_updated)
 }
 
+//For now get the DOT/currency exchange rate from coingecko API
+#[no_mangle]
+pub unsafe extern "C" fn update_market_data_xt(
+	genesis_hash: *const u8,
+	genesis_hash_size: u32,
+	nonce: &mut u32,
+	currency: *const u8,
+	currency_size: u32,
+	unchecked_extrinsic: *mut u8,
+	unchecked_extrinsic_size: u32,
+) -> sgx_status_t {
+	let genesis_hash_slice = slice::from_raw_parts(genesis_hash, genesis_hash_size as usize);
+	let genesis_hash = hash_from_slice(genesis_hash_slice);
+
+	let mut curr_slice = slice::from_raw_parts(currency, currency_size as usize);
+	let curr: String = Decode::decode(&mut curr_slice).unwrap();
+
+	//For now polkadot
+	let coin = "polkadot";
+
+	//Get the exchange rate
+	let url = CoinGeckoClient::base_url().unwrap();
+	let mut coingecko_client = CoinGeckoClient::new(url);
+	let rate = match coingecko_client.get_exchange_rate(coin, &curr) {
+		Ok(r) => r,
+		Err(x) => {
+			error!("[-] Failed to get the newest exchange rate from coingecko. {:?}", x);
+			return sgx_status_t::SGX_ERROR_UNEXPECTED
+		},
+	};
+	let extrinsic_slice =
+		slice::from_raw_parts_mut(unchecked_extrinsic, unchecked_extrinsic_size as usize);
+
+	let signer = Ed25519Seal::unseal().unwrap();
+	let call =
+		([TEERACLE_MODULE, UPDATE_EXCHANGE_RATE], curr.encode(), Some(U32F32::from_num(rate)));
+
+	let xt = compose_extrinsic_offline!(
+		signer,
+		call,
+		*nonce,
+		Era::Immortal,
+		genesis_hash,
+		genesis_hash,
+		RUNTIME_SPEC_VERSION,
+		RUNTIME_TRANSACTION_VERSION
+	)
+	.encode();
+
+	*nonce += 1;
```

Comment:
Composing the extrinsic (and updating the nonce) can be done with the extrinsics factory, after #498 is merged